src/pykeen/nn/emb.py

```diff
@@ -1068,3 +1070,103 @@ def forward(
         if indices is not None:
             x = x[indices]
         return x
+
+
+class NodePieceRepresentation(RepresentationModule):
+    """
+    Basic implementation of node piece decomposition [galkin2021]_.
+
+    .. note ::
+        This implementation currently only supports representation of entities by bag-of-relations.
+    """
+
+    #: the token representations
+    tokens: torch.FloatTensor
+
+    #: the entity-to-token mapping
+    assignment: torch.LongTensor
+
+    def __init__(
+        self,
+        *,
+        triples_factory: CoreTriplesFactory,
+        token_representation: EmbeddingSpecification,
+        aggregation: Callable[[torch.Tensor, int], torch.Tensor] = None,
+        k: int = 1,
+        **kwargs,
+    ):
+        """
+        Initialize the representation.
+
+        :param triples_factory:
+            the triples factory
+        :param token_representation:
+            the token representation specification
+        :param aggregation:
+            aggregation of multiple token representations to a single entity representation
+        :param k:
+            the number of tokens for each entity.
+        """
+        if triples_factory.create_inverse_triples:
+            raise NotImplementedError("Triples factories with inverse triples are currently not supported.")
+
+        # create token representations
+        tokens = token_representation.make(
```

Comment:
Can we go deeper? can token representations be any arbitrary RepresentationModule? or does it have to be embeddings?

Reply:
I think that would be *very* difficult

Reply:
We should allow it to pass an instantiated representation in case we want to build a model whose `__init__` takes care of instantiating something more exotic, like R-GCN representations!!

Reply:
Although, notice that the token representations are (at least in this variant) only used for *relation* representations.

Reply:
done in 70bf759

Reply:
because the arguments are potentially very different and need to be handled differently

I could be wrong though

Reply:
In the easiest case, we would not handle the parameters at all, but require the user to provide them properly.

```python
def __init__(
  self,
  ...,
  representation: HintOrType[Representations],
  representation_kwargs: Optional[Mapping[str, Any]] = None,
  ...,
):
  # ...
  representation = representation_resolver.make(representation, representation_kwargs)
```

Reply:
I think this is moot at the moment since the way these are constructed for the relation is now pretty specific.